contrib/migrate/migrate.c

```diff
@@ -0,0 +1,897 @@
+  "create index i__db_class_class_name_owner on _db_class (class_name, owner)",
+
+  "drop index i__db_class_class_name on _db_class"
+};
+
+/* only system classes update unique_name. */
+static char *update_db_class_for_system_classes =
+  "update _db_class set unique_name = class_name where is_system_class % 8 != 0";
+
+static void
+print_errmsg (const char *err_msg)
+{
+  fprintf (stderr, "ERROR: %s\n", err_msg);
+
+  return;
+}
+
+static void
+print_log (const char *log_fmt, ...)
+{
+  char log_msg[BUF_LEN];
+  time_t t;
+  struct tm *tm;
+  va_list ap;
+
+  t = time (NULL);
+  tm = localtime (&t);
+
+  snprintf (log_msg, BUF_LEN, "%d-%02d-%02d %02d:%02d:%02d\t",
+	    tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday, tm->tm_hour, tm->tm_min, tm->tm_sec);
+
+  va_start (ap, log_fmt);
+
+  vsnprintf (log_msg + strlen (log_msg), BUF_LEN - strlen (log_msg), log_fmt, ap);
+
+  va_end (ap);
+
+  snprintf (log_msg + strlen (log_msg), BUF_LEN - strlen (log_msg), "\n");
+
+  fprintf (stderr, "ERROR : %s\n", log_msg);
+
+  return;
+}
+
+static int
+char_isspace (int c)
+{
+  return ((c) == ' ' || (c) == '\t' || (c) == '\r' || (c) == '\n' || (c) == '\f' || (c) == '\v');
+}
+
+static char *
+next_char (char *str_p)
+{
+  char *p;
+
+  p = str_p;
+  while (char_isspace ((int) *p) && *p != '\0')
+    {
+      p++;
+    }
+
+  return (p);
+}
+
+static char *
+get_token (char *str_p, char **token_p)
+{
+  char *p, *end, *token = NULL;
+  int length;
+
+  token = NULL;
+  p = str_p;
+  while (char_isspace ((int) *p) && *p != '\0')
+    {
+      p++;
+    }
+  end = p;
+  while (!char_isspace ((int) *end) && *end != '\0')
+    {
+      end++;
+    }
+
+  length = (int) (end - p);
+  if (length > 0)
+    {
+      token = (char *) malloc (length + 1);
+      if (token != NULL)
+	{
+	  strncpy (token, p, length);
+	  token[length] = '\0';
+	}
+    }
+
+  *token_p = token;
+  return (end);
+}
+
+static int
+get_directory_path (char *buffer)
+{
+  const char *env_name;
+
+  env_name = getenv (DATABASES_ENVNAME);
+  if (env_name == NULL || strlen (env_name) == 0)
+    {
+      printf ("migrate: env variable %s is not set\n", DATABASES_ENVNAME);
+      return -1;
+    }
+  else
+    {
+      if (env_name[strlen (env_name) - 1] == '/')
+	{
+	  sprintf (buffer, "%s%s", env_name, DATABASES_FILENAME);
+	}
+      else
+	{
+	  sprintf (buffer, "%s/%s", env_name, DATABASES_FILENAME);
+	}
+    }
+
+  return 0;
+}
+
+static int
+get_db_path (char *dbname, char **pathname)
+{
+  FILE *file_p = NULL;
+  char line[MAX_LINE];
+  char filename[PATH_MAX];
+  char *vol_path = NULL;
+  char *name = NULL;
+  char *host = NULL;
+  char *str = NULL;
+
+  if (get_directory_path (filename) < 0)
+    {
+      return -1;
+    }
+
+  file_p = fopen (filename, "r");
+
+  while (fgets (line, MAX_LINE - 1, file_p) != NULL)
+    {
+      str = next_char (line);
+      if (*str != '\0' && *str != '#')
+	{
+	  str = get_token (str, &name);
+	  str = get_token (str, &vol_path);
```

Comment:
Users can set a log path that is different from the DB path.
We need to get "log-path", not "vol_path".